_source/_posts/2023-07-27-spring-graphql-react.md

```diff
@@ -0,0 +1,1794 @@
+        }
+      } catch (err) {
+        // Inactivity timeout
+        console.log("getAccessTokenSilently error", err);
+      }
+    }
+  }, [getAccessTokenSilently, isAuthenticated, accessToken]);
+
+  return {
+    saveAccessToken,
+  };
+};
+```
+
+The hook will call Auth0 `getAccessTokenSilently()` and will trigger a token refresh if the access token is expired. Then it will update axios interceptors to set the updated bearer token value in the request headers.
+Create the `useAsyncWithToken` hook:
+
+__useAsyncWithToken.tsx__
+```tsx
+import { useAccessToken } from "./useAccessToken";
+import { useAsync } from "react-use-custom-hooks";
+
+export const useAsyncWithToken = <T, P, E = string>(
+  asyncOperation: () => Promise<T>, deps: any[]
+) => {
+  const { saveAccessToken } = useAccessToken();
+  const [ data, loading, error ] = useAsync(async () => {
+    await saveAccessToken();
+    return asyncOperation();
+  }, {},  deps);
+
+  return {
+    data,
+    loading,
+    error
+  };
+};
+```
+
+Update the calls in the `CompanyTableContainer` component to use the `useAsyncWithToken` hook:
+
+__CompanyTableContainer.tsx__
+```tsx
+...
+const {
+  data: dataList,
+  loading: loadingList,
+  error: errorList,
+} = useAsyncWithToken(
+  () => CompanyApi.getCompanyList({ page: page - 1}),
+  [props.page]
+);
+
+const { data: dataCount } = useAsyncWithToken(
+  () => CompanyApi.getCompanyCount(),
+  []
+);
+...
+```
+
+Run the application with:
+
+```shell
+npm run dev
+```
+
+Go to http://localhost:3000 and you should be redirected to the Auth0 universal login page. After login in, you should see the companies list again.
+
+{% img blog/spring-graphql-react/auth0-universal-login.png alt:"Auth0 universal login form" width:"400" %}{: .center-image }
+
+{% img blog/spring-graphql-react/auth0-authorize-app.png alt:"Auth0 authorize application form" width:"400" %}{: .center-image }
+
+Once the companies load, you can inspect the network requests and see the bearer token is sent in the request headers. It will look like the example below:
+
+```
+Authorization: Bearer eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IlJzZHNsM211SjNYU2ZVT0tDOEMxSiJ9.eyJodHRwczovL3d3dy5qaGlwc3Rlci50ZWNoL3JvbGVzIjpbIlJPTEVfQURNSU4iLCJST0xFX1VTRVIiXSwiaXNzIjoiaHR0cHM6Ly9kZXYtYXZ1cDJsYXoudXMuYXV0aDAuY29tLyIsInN1YiI6ImF1dGgwfDY0MzQxOTkxNTJmYjc2N2Y3ZWFlZDU2NyIsImF1ZCI6WyJhcGktc2VydmVyIiwiaHR0cHM6Ly9kZXYtYXZ1cDJsYXoudXMuYXV0aDAuY29tL3VzZXJpbmZvIl0sImlhdCI6MTY5MTU5NDE3NCwiZXhwIjoxNjkxNTk0Mjk0LCJhenAiOiI1WW5QeEJiTjRoYXFyd2JNaXRpZFBRVTBjb0l2YUI3SCIsInNjb3BlIjoib3BlbmlkIHByb2ZpbGUgZW1haWwgb2ZmbGluZV9hY2Nlc3MifQ.oRMJNIwSeO9dfHi6Q2tr_B51YetEPZqjcVEEIBe3ky9tEe50tTB5ssMTbVaR78_1qXA663Cn4EMPEYTLlb_wiOHEqnKpZnpq0O07G2MGszkrgv5giaQBOXvr9UT_Kc3pFPK-xMVmOsicLoF_mz8iyOzReG1Gcw0UbS1fZsJthdtC9svLiPGC1rn-dwPJxPpKy118vWLbEgO3NsdVfaiPxbucv0TL_B8Msd-wBD8N4M_yej9jl7w2JT0ltLza_-Glbxr1aQBKd19O2QT8ovgyE325BYqeYiUOaV7efwMgqHUm7z4LngaPLRNhMLP2BExG7bUQD2JjH2Mh19LFYNM-Gw
```

Comment:
I like how this validates at jwt.io! :)

Reply:
Ooops, it shows the jhipster roles! I think I should make it invalid.

Reply:
I decided to just display the first characters.